display_manager.py

```diff
@@ -44,14 +57,105 @@ def gen_map(cls, player):
+        in a FIFO manner.
+
+        @param new_msg: str -> Message to display in the UI log, no size limit
+            for the string, and it will automatically be wrapped, however if
+            the string is too big for the console it might get clipped.
+        @param color: Colors -> Color of the text to be displayed, white by
+            default.
+
+        /!\ Attention, this method shouldn't be used through the DisplayManager
+        itself, instead use misc.add_message which is a standalone function /!\
+        """
+        wrapped = textwrap.wrap(new_msg, cls.MSG_WIDTH)
+
+        for line in wrapped:
+            if len(cls.game_msgs) == cls.MSG_HEIGHT:
+                del cls.game_msgs[0]
+
+            cls.game_msgs.append((line, color))
+
+    def _render_messages(cls):
+        """
+        Renders all messages found in the DisplayManager.game_msgs queue
+        """
+        for y, msg in enumerate(cls.game_msgs):
+            line, color = msg
+            cls.panel.draw_str(cls.MSG_X, y + 1, line, color, None)
+
+    def add_bar(cls, x, y, total_w, name, val, maxi, fg_color, bg_color):
+        """
+        Adds a bar to the UI in a chosen color with chosen text.
+
+        Useful for displaying stuff like HP, MP, EXP and any other thing that
+        the user might want to track through the UI in a min-max model.
+
+        The bar's color depends on the ratio of val to maxi, effectively
+        creating a visual representation of the stat.
+
+        Note: By default, the overlaid text (name) is in Colors.WHITE, so if
+        the fg_color/bg_color is Colors.WHITE, the text will seem invisible
+
+        @param x: int -> X coordinate relative to the container (panel console)
+        @param y: int -> Y coordinate relative to the container
+        @param total_w: int -> Total width of the bar, in pixels
+        @param name: str -> Name of the stat to track, to be displayed inside
+            the bar in a format such as {name}: {val}/{maxi}
+        @param val: int|float -> Current value of this stat
+            /!\ If using floats, verify the amount of decimal precision
+            since if the amount of precision is too big the display may be
+            affected in a negative manner /!\
+        @param maxi: int|float -> Max value of this stat, same warning as above
+        @param fg_color: Colors -> Color of the bar when "full".
+        @param bg_color: Colors -> Color of the bar when "empty".
+        """
+        bar_width = int(float(val) / maxi * total_w)
+        cls.panel.draw_rect(x, y, total_w, 1, None, bg=bg_color)
+
+        if bar_width > 0:
+            cls.panel.draw_rect(x, y, bar_width, 1, None, bg=fg_color)
+
+        # FIXME: make val be an int or a properly truncated float, don't coerce
+        text = f"{name}: {int(val)}/{maxi}"
+        x_centered = x + (total_w - len(text))//2
+        cls.panel.draw_str(x_centered, y, text, fg=Colors.WHITE, bg=None)
```

Comment:
You could add an optional text color argument.

Reply:
good idea